lib/resources/pam.rb

```diff
@@ -0,0 +1,370 @@
+    # @return [Boolean] true if found, false otherwise
+    def include?(rules, opts={:exact => false, :service_name => nil})
+      raise PamError, 'opts must be a hash' unless opts.is_a?(Hash)
+
+      service_name = get_service_name(opts[:service_name])
+
+      rules = Array(rules).map{|l| Pam::Rule.new(l, {:service_name => service_name})}
+
+      retval = false
+
+      if opts[:exact]
+        # This requires everything between the first and last rule to match
+        # exactly
+
+        first_entry = index(rules.first)
+        last_entry = index(rules.last)
+
+        if first_entry && last_entry
+          retval = (self[first_entry..last_entry] == rules)
+        end
+      else
+        # This match allows other rules between the two in question
+        retval = (rules.select{|l| super(l)} == rules)
+      end
+
+      return retval
+    end
+    alias_method :match, :include?
+
+    # An alias for setting `:exact => true` in the `include` method
+    def include_exactly?(rules, opts={})
+      include?(rules, opts.merge({:exact => true}))
+    end
+    alias_method :match_exactly, :include_exactly?
+
+    # Convert the data structure to an Array suitable for an RSpec diff
+    #
+    # @return [Array[String]]
+    def to_a
+      self.sort_by{|l| l.type}.map{|l| l.to_s}
+    end
+
+    # Convert the data structure to a String
+    #
+    # @return [String]
+    def to_s
+      to_a.join("\n")
+    end
+
+    private
+
+    # Get the service name out of the configuration target
+    #
+    # @param [String] svc_name Optional name of the service that should be
+    #    returned
+    #
+    # @return String
+    def get_service_name(svc_name = nil)
+      return svc_name if svc_name
+
+      if !svc_name && @config_target.directory?
+        raise PamError, 'You must pass ":service_name" as an option!'
+      else
+        return @config_target.basename
+      end
+    end
+  end
+
+  # A single Rule object that has been processed
+  #
+  # Rule equality is a fuzzy match that can accept regular expression matches
+  # within the string to compare
+  class Rule
+    attr_reader :to_s
+    attr_reader :service, :silent, :type, :control, :module_path, :module_arguments
+
+    def initialize(rule, opts = {})
+      @to_s = rule.strip.gsub(/\s+/,' ')
+
+      rule_regex = <<-'EOM'
+        # Start of Rule
+          ^
+        # Ignore initial Whitespace
+          \s*
+        # Capture Silent Flag
+          (?<silent>-)?
+      EOM
+
+      unless opts[:service_name]
+        rule_regex += <<-'EOM'
+          # Capture Service
+            (?<service_name>.+?)\s+
+        EOM
+      end
+
+      rule_regex += <<-'EOM'
+        # Capture Type
+          (?<type>.+?)\s+
+        # Capture Control
+          (?<control>(\[.+\]|.+?))\s+
+        # Capture Module Path
+          (?<module_path>.+?(\.so)?)
+        # Capture Module Args
+          (\s+(?<module_args>.+?))?
+        # End of Rule
+          $
+      EOM
```

Comment:
```ruby
# original string-based regexp
rule_regex = <<-'EOM'
           # Start of Rule
             ^
           # Ignore initial Whitespace
             \s*
           # Capture Silent Flag
             (?<silent>-)?
         EOM

         unless opts[:service_name]
           rule_regex += <<-'EOM'
             # Capture Service
               (?<service_name>.+?)\s+
           EOM
         end

         rule_regex += <<-'EOM'
           # Capture Type
             (?<type>.+?)\s+
           # Capture Control
             (?<control>(\[.+\]|.+?))\s+
           # Capture Module Path
             (?<module_path>.+?(\.so)?)
           # Capture Module Args
             (\s+(?<module_args>.+?))?
           # End of Rule
             $
         EOM
```
versus:
```ruby
# regexp with /x flag & interpolation:
%r%
  # Start of Rule
    ^
  # Ignore initial Whitespace
    \s*
  # Capture Silent Flag
    (?<silent>-)?
  # Capture Service
    #{'(?<service_name>.+?)\s+' unless opts[:service_name]}
  # Capture Type
    (?<type>.+?)\s+
  # Capture Control
    (?<control>(\[.+\]|.+?))\s+
  # Capture Module Path
    (?<module_path>.+?(\.so)?)
  # Capture Module Args
    (\s+(?<module_args>.+?))?
  # End of Rule
    $
%x
```
versus:
```ruby
# cleaned up:
sn = opts[:service_name]
%r%
    ^                                      # Start of Rule
    \s*                                    # Ignore initial Whitespace
    (?<silent>-)?                          # Capture Silent Flag
    #{'(?<service_name>.+?)\s+' unless sn} # Capture Service
    (?<type>.+?)\s+                        # Capture Type
    (?<control>(\[.+\]|.+?))\s+            # Capture Control
    (?<module_path>.+?(\.so)?)             # Capture Module Path
    (\s+(?<module_args>.+?))?              # Capture Module Args
    $                                      # End of Rule
%x
```

Now that the comments aren't being noise, I can start to pattern match potential problems. Should some of those `.+?` fragments be `\S+` or are they allowed to have whitespace in them? Do you want to capture that last `\s+` in the `module_args` bit? It could be `\s*` on the tail of the previous to be constent but still compatible w/ the last `?`.

Reply:
Honestly, I'll probably have to test this one some more to be able to answer all of your questions. I'll start with it as-is and then it can be refined over time.

Thanks for compressing the regex though, that was helpful and I'll update it accordingly.